File: CSE-216/hw4/BinarySearchTree.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class BinarySearchTree<T extends Comparable<T>> implements Iterable<T>
{
    private BTNode<T> root;
    private String name;

    public BinarySearchTree(String name)
    {
        root = null;
        this.name = name;
    }

    public boolean isEmpty()
    {
        return root == null;
    }

    public void add(T element)
    {
        if(isEmpty())
        {
            root = new BTNode<T>(element);
        }
        else
        {
            addHelp(root, element);
        }
    }

    private void addHelp(BTNode<T> node, T element)
    {
        if(element.compareTo(node.data) > 0)
        {
            if(node.right != null)
            {
                addHelp(node.right, element);
            }
            else 
            {
                BTNode<T> val = new BTNode<>(element);
                node.setRight(val);
            }
        }
        else if(element.compareTo(node.data) < 0)
        {
            if(node.left != null)
            {
                addHelp(node.left, element);
            }
            else
            {
                BTNode<T> val = new BTNode<>(element);
                node.setLeft(val);
            }
        }
    }

    public void addAll(List<T> elements)
    {
        elements.forEach(i -> add(i));
    }

    @Override
    public Iterator<T> iterator() 
    {
        List<T> order = new ArrayList<>();
        iteratorHelp(root, order);
        return order.listIterator();
    }

    public void iteratorHelp(BTNode<T> node, List<T> order)
    {
        if(node.left != null)
        {
            iteratorHelp(node.left, order);
        }

        order.add(node.data);

        if(node.right != null)
        {
            iteratorHelp(node.right, order);
        }
    }

    public String toString()
    {
        String display = "[" + name + "]";
        display += " " + toStringHelp(root,"");
        return display;
    }

    private String toStringHelp(BTNode<T> node, String display)
    {
        display += node.toString();
        
        if(node.left != null)
        
        {
            display += " L:(" + toStringHelp(node.left, "") + ")";
        }

        if(node.right != null)
        {
            display += " R:(" +  toStringHelp(node.right, "") + ")";
        }

        return display;
    }

    public class BTNode<E extends Comparable<E>> 
    {
        private E data;
        private BTNode<E> left;
        private BTNode<E> right;

        public BTNode(E data)
        {
            if(data == null)
            {
                throw new IllegalArgumentException("Invalid data entered");
            }

            this.data = data;
            this.left = null;
            this.right = null;
        }

        public void setLeft(BTNode<E> left)
        {
            this.left = left;
        }

        public void setRight(BTNode<E> right)
        {
            this.right = right;
        }
        
        public String toString()
        {
            return "" + data;
        } 
    }
}


    

File: CSE-216/hw4/binarysearchtree.py
class Node:
    def __init__(self):
        self.data = None
        self.left = None
        self.right = None

    def setLeft(self, left):
        self.left = left

    def setRight(self, right):
        self.right = right

    def setData(self, data):
        self.data = data

    def __str__(self):
        return str(self.data)
    
    def isEmpty(self):
        return self.data == None
    
    def __iter__(self):
        self.stack = [] 
        self.travel(self)
        return iter(self.stack)

    def travel(self, node):
        if(node.left):
            self.travel(node.left)

        self.stack.append(node.data)

        if(node.right):
            self.travel(node.right) 

    
class BinarySearchTree:
    def __init__(self, name:str, root:Node):
        self.root = root
        self.name = name
    
    def add(self, element):
        if(self.root.isEmpty()):
            self.root.setData(element)
        else:
            self.addHelp(self.root,element)


    def addHelp(self, node, element):
        if(element > node.data):
            if node.right is None:
                newnode = Node()
                newnode.setData(element)
                node.setRight(newnode)
            else:
                self.addHelp(node.right, element)

        elif(element < node.data):
            if node.left is None:
                newnode = Node()
                newnode.setData(element)
                node.setLeft(newnode)
            else:
                self.addHelp(node.left, element)

    def add_all(self, *val):
        for i in range(len(val)):
            self.add(val[i])

    def __str__(self) -> str:
        display = "[" + self.name + "]"
        display += " " + self.showHelp(self.root,"")
        return display

    def showHelp(self, node, display):
        display += node.__str__()

        if(node.left):
            display += " L:(" + self.showHelp(node.left, "") + ")"

        if(node.right):
            display += " R:(" + self.showHelp(node.right, "") + ")"

        return display
            

if __name__ == "__main__":
    tree = BinarySearchTree(name="Oak", root=Node())
    # the next line adds the elements in the order 5, 3, 9, and then 0 tree.add_all(5, 3, 9, 0)
    tree.add_all(5,7,9,0)
    print(tree)
    t1 = BinarySearchTree(name="Oak", root=Node()) 
    t1.add_all(1, 0, 10, 2, 7)
    for x in t1.root:
        print(x)


